lib/charms/tempo_coordinator_k8s/v0/tempo_api.py

```diff
@@ -0,0 +1,276 @@
+        application has not sent data.  To distinguish between these cases, check if
+        len(tempo_api_requirer.relations)==0.
+        """
+        relations = self.relations
+        if len(relations) == 0:
+            return None
+
+        # Being a little cautious here using getattr and get, since some funny things have happened with relation data
+        # in the past.
+        raw_data_dict = getattr(relations[0], "data", {}).get(relations[0].app)
+        if not raw_data_dict:
+            return None
+
+        # The data in the databag is in format [str: json-string-of-dict].  Expand out those nested JSON dicts so we
+        # have the full object
+        raw_data_dict = {k: json.loads(v) for k, v in raw_data_dict.items()}
+
+        return TempoApiAppData.model_validate(raw_data_dict)
+
+    def _validate_relation_metadata(self):
+        """Validate that the provided relation has the correct metadata for this endpoint."""
+        charm_root = Path(__file__).absolute().parents[4]
+        # Do not check charmcraft.yaml, just metadata.yaml.  metadata.yaml is the only file that will be present in a
+        # running charm, and this validation step fails during unit testing anyway so we'll never need to look at
+        # charmcraft.yaml
+        try:
+            yaml_raw = (charm_root / "metadata.yaml").read_text()
+            relation = yaml.safe_load(yaml_raw)["requires"][self._relation_name]
+        except KeyError as e:
+            raise ValueError(
+                f"Relation '{self._relation_name}' not found in metadata.yaml."
+            ) from e
+
+        if relation.get("limit", None) != 1:
+            raise ValueError(
+                "TempoApiRequirer is designed for relating to a single application and must be used with limit=1."
+            )
+
+
+class TempoApiProvider:
+    """The provider side of the tempo-api relation."""
+
+    def __init__(
+        self,
+        relation_mapping: RelationMapping,
+        relation_name: str,
+        app: Application,
+    ):
+        """Initialize the TempoApiProvider object.
+
+        This object is for serializing and sending data to a relation that uses the tempo-api interface - it does
+        not automatically observe any events for that relation.  It is up to the charm using this to call publish when
+        it is appropriate to do so, typically on at least the charm's leader_elected event and this relation's
+        relation_joined event.
+
+        Args:
+            relation_mapping: The RelationMapping of a charm (typically `self.model.relations` from within a charm
+                              object).
+            app: This application.
+            relation_name: The name of the wrapped relation.
+        """
+        self._charm_relation_mapping = relation_mapping
+        self._app = app
+        self._relation_name = relation_name
+
+    @property
+    def relations(self):
```

Comment:
```suggestion
    def relations(self) -> Tuple[ops.Relation, ...]:
```
?

Reply:
Used `def relations(self) -> List[Relation, ...]:` instead because the RelationMapping returns a list, but otherwise :+1: